content/blog/2022/05/verifying-image-signatures.md

```diff
@@ -0,0 +1,125 @@
+
+We need an image with a signature that we can verify. You can use [cosign](https://github.com/sigstore/cosign) to sign
+your images. For this example we'll use the image
+`ghcr.io/kubewarden/test-verify-image-signatures:signed` that was signed using a key. 
+
+We will use the following public key to verify the signature:
+
+```
+-----BEGIN PUBLIC KEY-----
+MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEALYYVLuPXXcKajo5meRcU+c6/iZe
+Dh+o3ZZs1puVW5FyS1cOusnLn209R0IOFL479vm1iFn+auB41JJZUxfc8g== 
+-----END PUBLIC KEY-----
+```
+
+Let's create a cluster-wide policy that will verify all images from the Kubewarden ghcr repository, and
+we'll use the previous public key for verification:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: verify-image-signatures
+spec:
+  module: https://github.com/kubewarden/verify-image-signatures/releases/download/v0.1.1/policy.wasm
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: true
+  settings:
+    signatures:
+        - image: "ghcr.io/kubewarden/*"
+          pubKeys:
+            - |
+              -----BEGIN PUBLIC KEY-----
+              MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEALYYVLuPXXcKajo5meRcU+c6/iZe
+              Dh+o3ZZs1puVW5FyS1cOusnLn209R0IOFL479vm1iFn+auB41JJZUxfc8g==
+              -----END PUBLIC KEY-----
+EOF
+````
+
+Wait for the policy to be active:
+
+```
+kubectl wait --for=condition=PolicyActive clusteradmissionpolicies verify-image-signatures
+```
+
+Verify we can create pods with containers that are signed:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  name: verify-image-valid
+spec:
+  containers:
+  - name: test-verify-image
+    image: ghcr.io/kubewarden/test-verify-image-signatures:signed
+EOF
+```
+
+Then check that the image was modified with the digest
+
+```
+kubectl get pod verify-image-valid -o=jsonpath='{.spec.containers[0].image}'
+```
+
+```
+ghcr.io/kubewarden/test-verify-image-signatures:signed@sha256:1d9d3da4c60d27b77bb96bba738319c1c4424853fdd10f65982f9f2ca2422a72
+```
+
+Finally, let's try to create a pod with an image that it is not signed:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  name: verify-image-invalid
+spec:
+  containers:
+  - name: test-verify-image
+    image: ghcr.io/kubewarden/test-verify-image-signatures:unsigned
+EOF
+```
+
+We will get the following error:
+
+```
+Error from server: error when creating "STDIN": admission webhook "clusterwide-verify-image-signatures.kubewarden.admission" denied the request: Pod verify-image-invalid is not accepted: verification of image ghcr.io/kubewarden/test-verify-image-signatures:unsigned failed: error invoking wapc verify: Error(HostError("Callback evaluation failure: signatures can't be fetched for image: ghcr.io/kubewarden/test-verify-image-signatures:unsigned"))
```

Comment:
argh, this error is pretty awful. I thought we were showing "no signatures found" now..

Reply:
we can create a PR and change it to:

```
Error from server: error when creating "STDIN": admission webhook "clusterwide-verify-image-signatures.kubewarden.admission" denied the request: Pod verify-image-invalid is not accepted: verification of image ghcr.io/kubewarden/test-verify-image-signatures:unsigned failed: error invoking wapc verify: Error(HostError("Callback evaluation failure: no signatures found for image: ghcr.io/kubewarden/test-verify-image-signatures:unsigned"))
```

Reply:
The policy is rejecting the request and setting the rejection failure to be the error message we got from the the host.

Can we ensure only the actual message is returned to the user? Right now it looks like the `Debug` (as in `"{:?}"` ) trait of the `Error` is used.

It would be better to simply have:
```
"Callback evaluation failure: no signatures found for image: ghcr.io/kubewarden/test-verify-image-signatures:unsigned"
```

Reply:
https://github.com/kubewarden/policy-fetcher/issues/85 created

Reply:
PR for consuming the fix in the policy: https://github.com/kubewarden/verify-image-signatures/pull/12